### ISSUE TITLE
Setup Continous Deployment

### ISSUE THREAD

**Issue by csehlke** [Rostlab/JS18_ProjectA_Group2#8]
Setup continous deployment of our develop branch

**Comment by syxolk** [comment]
CI is in #6

**Comment by syxolk** [comment]
I'm not going to deploy this thing. It makes everything so much more complicated just because you wanted to split it in an API and UI part for no particular reason:
- I cannot use the default web app stuff from Azure because it doesn't do `npm install` on your subfolder and you can't configure that.
- I cannot just start two web apps because (apparently) you need this proxy thing that redirects `/api` -> `localhost:3001`.
- What I could do: Use the docker apps from Azure, setup a Dockerfile and a Container registry, build a container image inside travis, push that image to Azure (which is totally ridiculous). But: Two server processes need to be run in a single container at the same time which is not a good idea.

I require you two make the following changes:
- Merge your two `package.json` into one at the top level.
- Change your start script so that a single `npm start` at the top level starts everything

**Comment by jyotirmay123** [comment]
Sorry, We are not going to change anything now. Sometime you say, you did a great job by splitting it to two parts and sometimes just because you are not able to do something, that becomes bad!! Anyway, we are coming up with third engine, i.e. our in-house NLP engine trained with custom data generated by us and you need to deploy that as well on python server. You cannot say we cannot use it, because, I asked Cheng and he is okay with use of that. So it's your problem, you need to find out a solution to deploy multiple server in a fashion such that they can communicate with each other and I think micro-services are built for that. Please, try to find a solution for your problem, I am sure there must be something to handle this.  Still, I will try to merge both of them, but I can not guarantee anything now.

**Comment by jyotirmay123** [comment]
@syxolk By the way, I did the integration and push it to merge with develop. Take care of it. I also configure to run python based nlp engine from node js. So now, I hope, it won't make any issue for you to deploy everything. For nlp engine, you need to install everything manually before starting the nlp server, but anyway, that is not for now, I did not push it now. Just for your knowledge.  Let me know, if any issue.

**Comment by syxolk** [comment]
> Sometime you say, you did a great job by splitting it to two parts and sometimes just because you are not able to do something, that becomes bad!!

Dividing the architecture into a frontend and a backend doesn't mean you need two separate servers for that.
I never said I'm not able to do it. I just said it's going to be unnecessary work for me for no benefits on your side.

> Anyway, we are coming up with third engine

Communicate such important things earlier.

>  I did the integration and push it to merge with develop.

Thanks, I will take a look at it.

**Comment by jyotirmay123** [comment]
We are anyway not using two servers, but two different ports. or probably 3 ports in future two integrate nlp_engine.

**Comment by gyachdav** [comment]
Please message me on slack when this issue is resolved. Thanks!

**Comment by syxolk** [comment]
I don't have time today anymore but I will do it tomorrow. Seems like I have to setup Dockerfiles and Azure Container Service for this.

**Comment by syxolk** [comment]
Build and push image:
```
docker build -t group2 .
docker login --username js2018group2 js2018group2.azurecr.io
docker tag group2 js2018group2.azurecr.io/group2:latest
docker push js2018group2.azurecr.io/group2:latest
```

Current problem: Image is probably too large to be built and pushed by travis (ca. 1 GB).
Ideas:
- Use a common base image that includes Node.js and Python already.
- Use the 'Continuous Delivery' feature of Azure

Public demo: https://js2018-group2.azurewebsites.net/ (not yet working)

Another problem:
```
2018-03-12T00:36:39.749750224Z [2] 
2018-03-12T00:36:39.749802323Z [2] > igraph@2.0.0 start:nlp /app
2018-03-12T00:36:39.749810023Z [2] > node nlp.server.js
2018-03-12T00:36:39.749816723Z [2] 
2018-03-12T00:36:39.868685586Z [0] 
2018-03-12T00:36:39.868724186Z [0] > igraph@2.0.0 server /app
2018-03-12T00:36:39.868731786Z [0] > nodemon server/server.js
2018-03-12T00:36:39.868738386Z [0] 
2018-03-12T00:36:39.914132775Z [1] 
2018-03-12T00:36:39.914170875Z [1] > igraph@2.0.0 ui /app
2018-03-12T00:36:39.914190974Z [1] > ng serve -H 0.0.0.0 --proxy-config UI/proxy.conf.json
2018-03-12T00:36:39.914198574Z [1] 
2018-03-12T00:36:52.048924304Z [0] [33m[nodemon] 1.17.1[39m
2018-03-12T00:36:52.053449353Z [0] [33m[nodemon] to restart at any time, enter `rs`[39m
2018-03-12T00:36:52.053498053Z [0] [33m[nodemon] watching: *.*[39m
2018-03-12T00:36:52.053508052Z [0] [32m[nodemon] starting `node server/server.js`[39m
2018-03-12T00:37:06.402248981Z [0] in NodeIndexService
2018-03-12T00:37:06.402292480Z [0] in NodeIndexRoute
2018-03-12T00:37:08.213794507Z [0] Server listening at 127.0.0.1:3001
2018-03-12T00:37:41.280661615Z [1] Port 3001 is already in use. Use '--port' to specify a different port.
```

**Comment by syxolk** [comment]
Fixed the 'port in use' issue. Now express answers with:
```
Invalid Host header
```

I guess `ng serve` disallows non-localhost by default. It's also not designed to be used in production, so do you have an alternative?

**Comment by jyotirmay123** [comment]
Ohkay.. we will try some different alternative if possible. And we also plann to move out nlp engine out of the repo as this is actually not part of the actual project. Can we deploy it separately or if you can provide us the detail to deploy, we can also give it a try as we know the structure and installations better about our project. Else we can have a call to fix it together.
Let us know, whichever is feasible.

**Comment by syxolk** [comment]
After this hack 7a16c3ef2453087428de6ded1b5c01a68dbda344 , manual deployment works now: https://js2018-group2.azurewebsites.net/

However, we still have two problems:
- `ng server --disable-host-check` is not a good idea on public servers
- the docker image is super large

Also the 'Continuous Delivery' feature of Azure doesn't like to deploy it.

**Comment by csehlke** [comment]
So what are we doing with that now, is there a plan to fix this issue?

**Comment by syxolk** [comment]
I'm going to build an intermediate docker image that will be used to build the final image. That way the build time should be reduced quite a bit.

**Comment by romikps** [comment]
Please close the issues if it's finished. Thanks

**Comment by syxolk** [comment]
Can you please decide on how this project is supposed to be setup?!

This is the history of wtf:
1. You copied the whole RasaNLU to the repository.
2. You partly removed RasaNLU.
3. You tell me to just call `pip install rasa_nlu`.
4. You totally remove RasaNLU.
5. You are using a custom version of RasaNLU.
6. You are using a custom version of RasaNLU and you need `pip install rasa_nlu`.

**Comment by syxolk** [comment]
RasaNLU returns this error now:
```
{
   "name":"StatusCodeError",
   "statusCode":500,
   "message":"500 - \"{\\n  \\\"error\\\": \\\"unsupported pickle protocol: 3\\\"\\n}\"",
   "error":"{\n  \"error\": \"unsupported pickle protocol: 3\"\n}",
   "options":{
      "url":"http://localhost:5000/parse",
      "qs":{
         "q":"Make a scatter plot with age and pay rate.",
         "project":"igraph_nlp",
         "model":"model_20180313-225005"
      },
      "simple":true,
      "resolveWithFullResponse":false,
      "transform2xxOnly":false
   },
   "response":{
      "statusCode":500,
      "body":"{\n  \"error\": \"unsupported pickle protocol: 3\"\n}",
      "headers":{
         "transfer-encoding":"chunked",
         "date":"Sun, 08 Apr 2018 16:51:28 GMT",
         "connection":"close",
         "content-type":"application/json",
         "server":"TwistedWeb/17.9.0"
      },
      "request":{
         "uri":{
            "protocol":"http:",
            "slashes":true,
            "auth":null,
            "host":"localhost:5000",
            "port":"5000",
            "hostname":"localhost",
            "hash":null,
            "search":"?q=Make%20a%20scatter%20plot%20with%20age%20and%20pay%20rate.&project=igraph_nlp&model=model_20180313-225005",
            "query":"q=Make%20a%20scatter%20plot%20with%20age%20and%20pay%20rate.&project=igraph_nlp&model=model_20180313-225005",
            "pathname":"/parse",
            "path":"/parse?q=Make%20a%20scatter%20plot%20with%20age%20and%20pay%20rate.&project=igraph_nlp&model=model_20180313-225005",
            "href":"http://localhost:5000/parse?q=Make%20a%20scatter%20plot%20with%20age%20and%20pay%20rate.&project=igraph_nlp&model=model_20180313-225005"
         },
         "method":"GET",
         "headers":{

         }
      }
   }
}
```

Python version in the container: `python --version: 2.7.12`

**Comment by syxolk** [comment]
Can you help me?

**Comment by csehlke** [comment]
I'm sorry i didn't tell you earlier, but i just found out, that our model was saved using python 3.6 therefore the pickle protocol 3 is necessary to use our model

We'll update the readme soon. Sorry for the inconveniences😌

**Comment by jyotirmay123** [comment]
Hi @Hans, We may have some unmanageable issue with our dataset-upload feature as the underline module does not support or unreliable on uploading huge dataset. So if uploading does not work. Please use db dump file attached @
[here](https://github.com/Rostlab/JS18_ProjectA_Group2/files/1896858/sql.zip) to create sample dataset tables and config table data dump for plot testing.
Core_dataset will work but Indicator set is huge. So we truncated that dataset and loaded into db.

**Comment by syxolk** [comment]
Hmm, I don't have any alternative upload possibility so it will fail for me, too.

**Comment by jyotirmay123** [comment]
Yeah. In that case, Please use datadump.sql file to setup the database to run natural language queries on sample dataset like core and Indicators.
